### PR TITLE
CUMULUS-260 update pdr stats

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-260: PDR page on dashboard only shows zero](https://bugs.earthdata.nasa.gov/browse/CUMULUS-260)

Currently the pdr status stays in 'running' even if the granules are processed and completed.

## Changes

* Update lambda pdr-status-check to take pdr as additional input and output playload, so that pdr stats (granule stats in the pdr) can be updated by the cumulus api when the workflow is complete.
* Add lambda function sf-sns-report which is copied from @cumulus/api/lambdas/sf-sns-broadcase with modification, sf-sns-report can be used to report step function status anywhere inside a step function.  So add step sf-sns-report after each pdr-status-check, we will get the PDR status progress at real time.
* Update pdr-status-check to handle ExecutionDoesNotExist error.
* Fixed @cumulus/common/aws setGranuleStatus

## Test Plan
Things that should succeed before merging.

- [x] Unit tests - updated
- [x] Adhoc testing - tested in aws

Reviewers: @scisco @laurenfrederick

### CODE REVIEW

**Comment by jennyhliu** [comment]
Actually, we can call indexer.partialRecordUpdate to update the pdr.stats only from pdr-status-check.  I will try that

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 
+- **CUMULUS-260: "PDR page on dashboard only shows zeros."** The PDR stats in LPDAAC are all 0s, even if the dashboard has been fixed to retrieve the correct fields.  The current version of pdr-status-check has a few issues.
+  - pdr is not included in the input/output schema.  It's available from the input event.  So the pdr status and stats are not updated when the ParsePdr workflow is complete.  Adding the pdr to the input/output of the task will fix this.
+  - pdr-status-check doesn't update pdr stats which prevent the real time pdr progress from showing up in the dashboard. To solve this, added lambda function sf-sns-report which is copied from @cumulus/api/lambdas/sf-sns-broadcase with modification, sf-sns-report can be used to report step function status anywhere inside a step function.  So add step sf-sns-report after each pdr-status-check, we will get the PDR status progress at real time.
+  - It's possible a execution is still in the queue and doesn't exist in sfn yet.  Added code to handle 'ExecutionDoesNotExist' error when checking the execution status.
```
Comment (abarciauskas-bgse) [inline CHANGELOG.md]: minor grammar fix: `It's possible an execution is still in the queue`

File: cumulus/tasks/pdr-status-check/index.js [hunk 2]
```diff
@@ -137,28 +139,36 @@ function buildOutput(event, groupedExecutions) {
...
-    .then(groupExecutionsByStatus)
-    .then((groupedExecutions) => {
-      const counter = getCounterFromEvent(event) + 1;
-      const exceededLimit = counter >= getLimitFromEvent(event);
+  const executions = [];
+  for (const executionArn of runningExecutionArns) {
+    try {
+      const execution = await aws.sfn().describeExecution({ executionArn }).promise();
+      executions.push(execution);
+    }
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/pdr-status-check/index.js]: nit: should `} catch (e) {` should be on one line

File: cumulus/tasks/pdr-status-check/index.js [hunk 3]
```diff
@@ -137,28 +139,36 @@ function buildOutput(event, groupedExecutions) {
...
+    catch (e) {
+      // it's ok if a execution is still in the queue and has not be executed
+      if (e.code === 'ExecutionDoesNotExist') {
+        executions.push({ executionArn: executionArn, status: 'RUNNING' });
+      }
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/pdr-status-check/index.js]: Same thing here
```javascript
} else {
  throw e
}
```

File: cumulus/tasks/sf-sns-report/index.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    if (typeof event.exception === 'object') {
+      // this is needed to avoid flagging cases like "exception: {}" or "exception: 'none'"
+      if (Object.keys(event.exception).length > 0) {
+        return true;
+      }
+    }
+  }
+  // Error and error keys are not part of the cumulus message
+  // and if they appear in the message something is seriously wrong
+  else if (event.Error || event.error) {
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/index.js]: Do we know an instance of this occurring? A better option might be to whitelist the top-level keys we expect from a cumulus message, otherwise this seems unnecessary

File: cumulus/tasks/sf-sns-report/index.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+const cumulusMessageAdapter = require('@cumulus/cumulus-message-adapter-js');
+
+/**
+ * Determines if there was a valid exception in the input message
+ *
+ * @param {Object} event - aws event object
+ * @returns {boolean} true if there was an exception, false otherwise
+ */
+function eventFailed(event) {
+  if (event.exception) {
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/index.js]: Maybe flatten these nested conditionals?
```javascript
if (event.exception && typeof(event.exception) === 'object' && Object.keys(event.exception.length) > 0) {
  return true
}
```

File: cumulus/tasks/sf-sns-report/index.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  // and if they appear in the message something is seriously wrong
+  else if (event.Error || event.error) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * if the cumulus message shows that a previous step failed,
+ * this function extract the error message from the cumulus message
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/index.js]: `this function extracts the error message`

File: cumulus/tasks/sf-sns-report/index.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+ * and fail the function with that information. This ensures that the
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/index.js]: `and fails the function`

File: cumulus/tasks/sf-sns-report/index.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Lambda handler. It broadcasts an incoming Cumulus message to SNS
+ *
+ * @param {Object} event - a Cumulus Message
+ * @param {Object} context - an AWS Lambda context object
+ * @param {Function} callback - an AWS Lambda call back
+ * @returns {Promise} updated event object
+ */
+function handler(event, context, callback) {
+  return StepFunction.pullEvent(event).then((message) =>
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/index.js]: nit: wrap in curly brace:
```javascript
return StepFunction.pullEvent(event).then((message) => {
  cumulusMessageAdapter.runCumulusTask(publishSnsMessage, message, context, callback));
}
```

File: cumulus/tasks/sf-sns-report/tests/index.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  };
+
+  return publishSnsMessage(cloneDeep(event))
+    .catch((e) => {
+      t.is(e.message, event.input.error.Cause);
+    });
+});
+
+
+test('send report when sfn is finished and granule succeed', async (t) => {
```
Comment (abarciauskas-bgse) [inline cumulus/tasks/sf-sns-report/tests/index.js]: `send report when sfn is finished and granule has succeeded`

**Comment by abarciauskas-bgse** [comment]
@jennyhliu you mentioned in the PR description `Add lambda function sf-sns-report which is copied from /lambdas/sf-sns-broadcase with modification` should some of the shared functionality be pulled into a separate package which becomes a dependency of both to reduce duplication of code?

**Comment by scisco** [comment]
@abarciauskas-bgse I think we should retire that other lambda function in the api package. But retiring it will break existing deployments. so we should come up with a plan for that.

File: CHANGELOG.md (hunk 1 above)
Comment (jennyhliu) [reply CHANGELOG.md]: done

File: cumulus/tasks/pdr-status-check/index.js (hunk 2 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/index.js]: eslint doesn't like the one-line style

File: cumulus/tasks/pdr-status-check/index.js (hunk 3 above)
Comment (scisco) [reply cumulus/tasks/pdr-status-check/index.js]: yeah, our eslint doesn't like that.

File: cumulus/tasks/sf-sns-report/index.js (hunk 4 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: I don't.  I don't know if removing this logic will break something.

Comment (scisco) [reply cumulus/tasks/sf-sns-report/index.js]: this was reported in an earlier bug report and this is the fix for that bug.

File: cumulus/tasks/pdr-status-check/package.json [hunk 10]
```diff
@@ -40,6 +40,7 @@
     ]
   },
   "dependencies": {
+    "@cumulus/api": "^1.1.1",
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/package.json]: removed

File: cumulus/tasks/sf-sns-report/README.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
+# @cumulus/queue-pdrs
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/README.md]: corrected

File: cumulus/tasks/sf-sns-report/index.js (hunk 4 above)
Comment (abarciauskas-bgse) [reply cumulus/tasks/sf-sns-report/index.js]: ah ok - do you recall if that bug was before or after cumulus message adapter integration?

File: cumulus/tasks/pdr-status-check/index.js (hunk 2 above)
Comment (abarciauskas-bgse) [reply cumulus/tasks/pdr-status-check/index.js]: weird that seems inconsistent with https://github.com/airbnb/javascript#blocks--cuddled-elses

File: cumulus/tasks/sf-sns-report/index.js (hunk 5 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: yes.  this is neat

File: cumulus/tasks/sf-sns-report/index.js (hunk 6 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: done

File: cumulus/tasks/sf-sns-report/index.js (hunk 7 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: done

File: cumulus/tasks/sf-sns-report/index.js (hunk 8 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: done

File: cumulus/tasks/sf-sns-report/tests/index.js (hunk 9 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/tests/index.js]: done

**Comment by jennyhliu** [comment]
@abarciauskas-bgse @scisco updated based on comments.

**Comment by scisco** [comment]
@laurenfrederick we should figure out our roll out strategy for the updated sns message task. Currently deployments use the version included in the cumulus-api but they should switch to use this one. We might want to issue a deprecation warning or something for example.

File: cumulus/tasks/pdr-status-check/index.js (hunk 2 above)
Comment (yjpa7145) [reply cumulus/tasks/pdr-status-check/index.js]: That's because of this override: https://github.com/cumulus-nasa/cumulus/blob/master/.eslintrc.json#L86.

Our eslintrc is requiring the "stroustrup" style for braces:  https://eslint.org/docs/rules/brace-style

File: cumulus/tasks/pdr-status-check/index.js [hunk 12]
```diff
@@ -137,28 +139,36 @@ function buildOutput(event, groupedExecutions) {
...
-  const promisedExecutionDescriptions = runningExecutionArns.map((executionArn) =>
-    aws.sfn().describeExecution({ executionArn }).promise());
-
-  return Promise.all(promisedExecutionDescriptions)
-    .then(groupExecutionsByStatus)
-    .then((groupedExecutions) => {
-      const counter = getCounterFromEvent(event) + 1;
-      const exceededLimit = counter >= getLimitFromEvent(event);
+  const executions = [];
+  for (const executionArn of runningExecutionArns) {
```
Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/index.js]: The style guide recommends avoiding iterators when possible.  Instead of a `for` loop here, consider using `map`.

http://airbnb.io/javascript/#iterators--nope

Because of the special case where the execution does not yet exist, I would probably create a `describeExecutionStatus` function and use that in the map.  Something like:

```
function describeExecutionStatus(executionArn) {
  return aws.sfn().describeExecution({ executionArn }).promise()
    .catch((e) => {
      if (e.code === 'ExecutionDoesNotExist') {
        return { executionArn: executionArn, status: 'RUNNING' };
      }
      throw e;
    });
}
```

With that function, you could then remove the loop and just say
```
const executions = Promise.all(promisedExecutionStatuses);
```

File: cumulus/tasks/pdr-status-check/schemas/input.json [hunk 13]
```diff
@@ -23,6 +23,14 @@
     },
     "counter": { "type": "integer" },
     "limit": { "type": "integer" },
-    "isFinished": { "type": "boolean" }
+    "isFinished": { "type": "boolean" },
+    "pdr": {
```
Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/schemas/input.json]: Please add a description.

Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/schemas/input.json]: Please make this property required.

File: cumulus/tasks/pdr-status-check/index.js (hunk 3 above)
Comment (yjpa7145) [reply cumulus/tasks/pdr-status-check/index.js]: Same comment as the other instance:

That's because of this override: https://github.com/cumulus-nasa/cumulus/blob/master/.eslintrc.json#L86.

Our eslintrc is requiring the "stroustrup" style for braces:  https://eslint.org/docs/rules/brace-style

File: cumulus/tasks/pdr-status-check/schemas/output.json [hunk 14]
```diff
@@ -23,6 +23,14 @@
     },
     "counter": { "type": "integer" },
     "limit": { "type": "integer" },
-    "isFinished": { "type": "boolean" }
+    "isFinished": { "type": "boolean" },
+    "pdr": {
```
Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/schemas/output.json]: Please add a comment and make this a required property.

File: cumulus/tasks/pdr-status-check/tests/index.js [hunk 15]
```diff
@@ -61,26 +64,35 @@ test('returns the correct results in the nominal case', (t) => {
     'arn:1': 'RUNNING',
     'arn:2': 'SUCCEEDED',
     'arn:3': 'FAILED',
-    'arn:4': 'ABORTED'
+    'arn:4': 'ABORTED',
+    'arn:7': null
+  };
+  const error = {
+    message: 'Execution Does Not Exist: arn',
```
Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/tests/index.js]: This is always going to set the message to `Execution Does Not Exist: arn`, not the actual execution arn.

File: cumulus/tasks/pdr-status-check/tests/index.js [hunk 16]
```diff
@@ -61,26 +64,35 @@ test('returns the correct results in the nominal case', (t) => {
...
   };
 
   const stubSfnClient = {
     describeExecution: ({ executionArn }) => ({
-      promise: () => Promise.resolve({
-        executionArn,
-        status: executionStatuses[executionArn]
-      })
+      promise: () => {
+        if (executionStatuses[executionArn] === null) return Promise.reject(error);
```
Comment (yjpa7145) [inline cumulus/tasks/pdr-status-check/tests/index.js]: This should be returning an actual Error instance, not a simple object.  Consider something like:

```
if (!executionStatuses[executionArn]) {
  const error = new Error(`Execution does not exist: ${executionArn}`);
  error.code = 'ExecutionDoesNotExist';
  Promise.reject(error);
}
```

File: cumulus/tasks/pdr-status-check/index.js (hunk 12 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/index.js]: Thank you!  I realized we should keep Promise.all after I started to look at CUMULUS-304.  I will do this under CUMULUS-304: Add AWS API throttling to pdr-status-check task.

File: cumulus/tasks/pdr-status-check/schemas/input.json (hunk 13 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/schemas/input.json]: added description and made it a required field.

File: cumulus/tasks/pdr-status-check/schemas/output.json (hunk 14 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/schemas/output.json]: done

File: cumulus/tasks/pdr-status-check/tests/index.js (hunk 16 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/tests/index.js]: Thanks.  Done

File: cumulus/tasks/pdr-status-check/tests/index.js (hunk 15 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/tests/index.js]: done

File: packages/common/aws.js [hunk 17]
```diff
@@ -594,5 +596,7 @@ exports.setGranuleStatus = async (
 ) => {
   const key = exports.getGranuleS3Params(granuleId, stack, bucket);
   const executionArn = exports.getExecutionArn(stateMachineArn, executionName);
-  await exports.s3().putObject(bucket, key, '', null, { executionArn, status }).promise();
+  const params = { Bucket: bucket, Key: key };
+  params.Metadata = { executionArn, status };
+  await exports.s3().putObject(params).promise();
```
Comment (yjpa7145) [inline packages/common/aws.js]: I've filed a ticket to take a look at this functionality:  https://bugs.earthdata.nasa.gov/browse/CUMULUS-420

File: CHANGELOG.md [hunk 18]
```diff
@@ -17,6 +17,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
   - The `@cumulus/api` endpoints for collections, providers and rules _only_ query DynamoDB, with the exception of LIST endpoints and the collections' GET endpoint.
 
+- **CUMULUS-260: "PDR page on dashboard only shows zeros."** The PDR stats in LPDAAC are all 0s, even if the dashboard has been fixed to retrieve the correct fields.  The current version of pdr-status-check has a few issues.
+  - pdr is not included in the input/output schema.  It's available from the input event.  So the pdr status and stats are not updated when the ParsePdr workflow is complete.  Adding the pdr to the input/output of the task will fix this.
+  - pdr-status-check doesn't update pdr stats which prevent the real time pdr progress from showing up in the dashboard. To solve this, added lambda function sf-sns-report which is copied from @cumulus/api/lambdas/sf-sns-broadcase with modification, sf-sns-report can be used to report step function status anywhere inside a step function.  So add step sf-sns-report after each pdr-status-check, we will get the PDR status progress at real time.
```
Comment (yjpa7145) [inline CHANGELOG.md]: "broadcast", not "broadcase"

File: cumulus/tasks/sf-sns-report/index.js [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * Determines if there was a valid exception in the input message
+ *
+ * @param {Object} event - aws event object
+ * @returns {boolean} true if there was an exception, false otherwise
+ */
+function eventFailed(event) {
+  // event has exception
+  // and it is needed to avoid flagging cases like "exception: {}" or "exception: 'none'"
+  if (event.exception && (typeof event.exception === 'object') &&
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: The first check for event.exception is not necessary.  If it isn't set, then `(typeof event.exception === 'object')` will be false.  Can just be:

```
function eventFailed(event) {
  // event has exception
  // and it is needed to avoid flagging cases like "exception: {}" or "exception: 'none'"
  if ((typeof event.exception === 'object') &&
      (Object.keys(event.exception).length > 0)) return true;

  // Error and error keys are not part of the cumulus message
  // and if they appear in the message something is seriously wrong
  if (event.Error || event.error) return true;

  return false;
}
```

Comment (yjpa7145) [reply cumulus/tasks/sf-sns-report/index.js]: This example is also formatted so it's a little easier to read.

File: cumulus/tasks/sf-sns-report/index.js [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * if the cumulus message shows that a previous step failed,
+ * this function extracts the error message from the cumulus message
+ * and fails the function with that information. This ensures that the
+ * Step Function workflow fails with the correct error info
+ *
+ * @param {Object} event - aws event object
+ * @returns {undefined} throws an error and does not return anything
+ */
+function makeLambdaFunctionFail(event) {
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: It's a little hard to follow what this function is doing.  It might be easier with something like:

```
function buildError(type, cause) {
  let ErrorClass;

  if (Object.keys(errors).includes(type)) ErrorClass = errors[type];
  else if (type === 'TypeError') ErrorClass = TypeError;
  else ErrorClass = Error;

  return new ErrorClass(cause);
}

function makeLambdaFunctionFail(event) {
  const error = event.exception || event.error;

  if (!error) throw new Error('Step Function failed for an unknown reason.');
  
  throw buildError(error.Error, error.Cause);
}
```

File: cumulus/tasks/sf-sns-report/index.js [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * Lambda handler. It broadcasts an incoming Cumulus message to SNS
+ *
+ * @param {Object} event - a Cumulus Message
+ * @param {Object} context - an AWS Lambda context object
+ * @param {Function} callback - an AWS Lambda call back
+ * @returns {Promise} updated event object
+ */
+function handler(event, context, callback) {
+  return StepFunction.pullEvent(event).then((message) => {
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: Pulling the event is handled by the message adapter and should not be required here.

File: cumulus/tasks/sf-sns-report/schemas/config.json [hunk 22]
```diff
@@ -0,0 +1,28 @@
+{
+  "title": "SfSnsReportConfig",
+  "description": "Describes the config used by the sf-sns-report task",
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/schemas/config.json]: I assume that all of these properties are required, so they should be marked as required here.

File: cumulus/tasks/sf-sns-report/index.js [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+ * @param {Object} event.config.sfnEnd - indicate if it's the last step of the step function
+ * @param {string} event.config.stack - the name of the deployment stack
+ * @param {string} event.config.bucket - S3 bucket
+ * @param {string} event.config.stateMachine - current state machine
+ * @param {string} event.config.executionTime - execution time
+ * @returns {Promise.<Object>} - AWS SNS response. see schemas/output.json for detailed output
+ * schema that is passed to the next task in the workflow
+ */
+async function publishSnsMessage(event) {
+  const config = get(event, 'config', []);
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: The config is an Object, not an Array.  That being said, message adapter will ensure that config is always set, so there's no need for a default value.

File: cumulus/tasks/sf-sns-report/index.js [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+  const message = get(event, 'input', []);
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: The input is an Object, not an Array.  That being said, message adapter will ensure that config is always set, so there's no need for a default value.

File: cumulus/tasks/sf-sns-report/tests/index.js [hunk 25]
```diff
@@ -0,0 +1,87 @@
+'use strict';
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/tests/index.js]: None of these tests are verifying that the correct message was sent to SNS.  They are only looking at the return value from the function.  Localstack has SNS support, please update the tests to verify that the correct messages are being sent to SNS.

File: cumulus/tasks/sf-sns-report/index.js (hunk 19 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: updated

File: cumulus/tasks/sf-sns-report/index.js (hunk 20 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: Thanks.  This makes the code much clear.  Sorry I didn't try to understand the code.

File: cumulus/tasks/sf-sns-report/schemas/config.json (hunk 22 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/schemas/config.json]: Based on the code, they are only needed when the task is the last step of the step function and a granule state is set.

File: cumulus/tasks/sf-sns-report/index.js (hunk 23 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: ok

File: cumulus/tasks/sf-sns-report/index.js [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+      TopicArn: topicArn,
+      Message: JSON.stringify(message)
+    }).promise();
+  }
+
+  if (failed) {
+    makeLambdaFunctionFail(message);
+  }
+
+  return message;
```
Comment (scisco) [inline cumulus/tasks/sf-sns-report/index.js]: @jennyhliu I think we should NOT return the whole message here. If the user forgets to include the `outputs` key in the cumulus config, the message ends up getting larger and larger.

Could you instead just return the output of the sns aws call?

File: cumulus/tasks/sf-sns-report/index.js (hunk 19 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: actually ``(event.exception)` check is need to take care of the case `event.exception=null`, otherwise `Object.keys(event.exception)` will throw exception.

File: cumulus/tasks/pdr-status-check/index.js (hunk 12 above)
Comment (scisco) [reply cumulus/tasks/pdr-status-check/index.js]: if we don’t use a for loop here we should consider using p-limit or we will quickly hit the step function api limit.

File: cumulus/tasks/sf-sns-report/index.js [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+
+  if (topicArn) {
+    // if this is the sns call at the end of the execution
+    if (finished) {
+      message.meta.status = failed ? 'failed' : 'completed';
+      const granuleId = get(message, 'meta.granuleId', null);
+      if (granuleId) {
+        await setGranuleStatus(
+          granuleId,
+          get(config, 'stack', null),
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: Updated the code and schema, there is no good way to prevent the user from misconfiguring the parameters.

File: cumulus/tasks/sf-sns-report/index.js (hunk 26 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/index.js]: Updated the code to return the output from sns response, and the README.md

File: cumulus/tasks/sf-sns-report/tests/index.js (hunk 25 above)
Comment (jennyhliu) [reply cumulus/tasks/sf-sns-report/tests/index.js]: Updated the function to return the sns response.  If the MessageId is returned, then the message is sent to SNS.  There is no easy way to retrieve the message sent to SNS.

File: cumulus/tasks/pdr-status-check/index.js (hunk 12 above)
Comment (jennyhliu) [reply cumulus/tasks/pdr-status-check/index.js]: @scisco  I will fix it under CUMULUS-304.

**Comment by jennyhliu** [comment]
@scisco @yjpa7145 made updates based on comments.  please review

File: cumulus/tasks/sf-sns-report/index.js (hunk 4 above)
Comment (yjpa7145) [reply cumulus/tasks/sf-sns-report/index.js]: @abarciauskas-bgse brings up a very good point.  When you're using the message adapter, the only keys that the simplified event are going to have are "config" and "input".  This is never going to trigger.

File: cumulus/tasks/sf-sns-report/index.js [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+  let response = {};
+  if (topicArn) {
+    // if this is the sns call at the end of the execution
+    if (finished) {
+      message.meta.status = failed ? 'failed' : 'completed';
+      const granuleId = get(message, 'meta.granuleId', null);
+      if (granuleId) {
+        await setGranuleStatus(
+          granuleId,
+          get(config, 'stack'),
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/index.js]: Because you aren't looking deep into an object, there is no need for `get()` here.  Simply saying `config.stack` accomplishes the same thing.  The same goes for the next few lines as well.

File: cumulus/tasks/sf-sns-report/schemas/config.json [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+      "type": "string"
+    },
+    "executionTime": {
+      "description": "execution time (from cumulus_meta.execution_name). Required when sfnEnd is true and granule status is reported.",
+      "type": "string"
+    }
+  },
+  "oneOf": [
+    {
+      "required": ["stack", "bucket", "stateMachine", "executionTime"]
```
Comment (yjpa7145) [inline cumulus/tasks/sf-sns-report/schemas/config.json]: stack, bucket, and stateMachine should always be required.  Should `executionTime` actually be `executionName`?  I don't see `executionTime` anywhere in the code, but I do see `executionName`.

**Comment by scisco** [comment]
@jennyhliu the input.json should accept any object. Basically whatever you pass to the task it will publish to sns.
